Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -106,7 +107,13 @@ public actor SwiftLanguageServer: ToolchainLanguageServer {
   // FIXME: ideally we wouldn't need separate management from a parent server in the same process.
   var documentManager: DocumentManager
 
-  var currentDiagnostics: [DocumentURI: [CachedDiagnostic]] = [:]
+  /// For each edited document, the last task that was triggered to send a `PublishDiagnosticsNotification`.
+  ///
+  /// This is used to cancel previous publish diagnostics tasks if an edit is made to a document.
+  ///
+  /// - Note: We only clear entries from the dictionary when a document is closed. The task that the document maps to
+  ///   might have finished. This isn't an issue since the tasks do not retain `self`.
```

Comment:
Yes, because if we wanted to clear it straight away, we would need to be careful with race conditions. We would get into issues `inFlightPublishDiagnosticsTasks[document]` get set to `nil` at the end of a task if the task that the dictionary maps to isn’t the task because another publish diagnostic task was started.

And since it doesn’t really hurt to keep them around, I decided to just define away the possibility of race conditions here by only clearing it on document close.

Reply:
If another was started then we would have expected the diagnostics to be cancelled, which is relied on in `documentDiagnostic` anyway. So that race doesn't seem like something that can actually happen to me.

It seems like it would be nice to know whether diagnostics are in flight, though I suppose we could still do that with this by checking the state of the task 🤷